tests/transport/test_transport_compliance.py

```diff
@@ -0,0 +1,153 @@
+    """A basic WSGI app that exposes methods for testing http requests."""
+    app = flask.Flask(__name__)
+
+    @app.route('/basic')
+    def index():
+        return 'Basic Content', http_client.OK, {'X-Test-Header': 'value'}
+
+    @app.route('/server_error')
+    def server_error():
+        return 'Error', http_client.INTERNAL_SERVER_ERROR
+
+    @app.route('/authorized')
+    def authorized():
+        """Echoes the request auth and user agent headers."""
+        return flask.jsonify({
+            'authorization': flask.request.headers.get('authorization'),
+            'user_agent': flask.request.headers.get('user-agent')
+        })
+
+    @app.route('/authorized_refresh')
+    def authorized_refresh():
+        """Returns UNAUTHORIZED if the request token doesn't contain
+        '-new', otherwise returns the auth and user agent headers."""
+        authorization = flask.request.headers.get('authorization', '')
+
+        if '-new' not in authorization:
+            return 'Not authorized', http_client.UNAUTHORIZED
+
+        return flask.jsonify({
+            'authorization': flask.request.headers.get('authorization'),
+            'user_agent': flask.request.headers.get('user-agent')
+        })
+
+    return app.wsgi_app
+
+
+class MockCredentials(object):
+    def __init__(self, token):
+        self.token = token
+        self.user_agent = 'test-credentials'
+
+    def _before_request(self, http, uri, headers):
+        headers[b'authorization'] = 'Bearer {}'.format(self.token)
+
+    def refresh(self, http):
+        self.token = '{}{}'.format(self.token, '-new')
+
+
+class TransportComplianceTests(object):
+    """Test base class to confirm that a transport satisfies the interface
+    expected by oauth2client."""
+    transport = None
+
+    @pytest.fixture(autouse=True)
+    def testserver(self):
+        """Provides a test HTTP server that is automatically created before
+        a test and destoryed at the end. This server is running the test
+        app defined above."""
+        server = WSGIServer(application=test_server_app())
+        server.start()
+        self.server = server
+        yield
+        server.stop()
+
+    def test_request_basic(self):
+        http = self.transport.get_http_object()
+        response = self.transport._request(
+            http, self.server.url + '/basic')
+
+        assert response.status == http_client.OK
+        print(response.headers.keys())
+        assert response.headers['x-test-header'] == 'value'
+        assert response.data == b'Basic Content'
+
+    def test_request_error(self):
+        http = self.transport.get_http_object()
+        response = self.transport._request(
+            http, self.server.url + '/server_error')
+
+        assert response.status == http_client.INTERNAL_SERVER_ERROR
+        assert response.data == b'Error'
+
+    def test_connection_error(self):
+        http = self.transport.get_http_object()
+
+        with pytest.raises(Exception):
```

Comment:
So that's actually an outstanding question. There's a lot of errors that a transport can raise. Should we force the transports to raise a consistent exception? oauth2client basically never tried to catch any transport-level exceptions- preferring instead to let them work up the stack and have the user deal with them.